src/wms/src/main/java/org/geoserver/wms/legendgraphic/LegendGraphicBuilder.java

```diff
@@ -153,8 +153,8 @@ protected Style resizeForDPI(GetLegendGraphicRequest request, Style gt2Style) {
         double standardDpi = RendererUtilities.getDpi(Collections.emptyMap());
         if (dpi != standardDpi) {
             double scaleFactor = dpi / standardDpi;
-            w = ((int) Math.round(w * scaleFactor));
-            h = ((int) Math.round(h * scaleFactor));
+            w = ((int) Math.round(request.getWidth() * scaleFactor));
+            h = ((int) Math.round(request.getHeight() * scaleFactor));
```

Comment:
I am not familiar with this bit of code, but after review, the fix seems legit.
However, I'm not convinced it's the best way to solve this. The method is called over and over, once for each layer, the reason why we're having this problem it's that the code does two separate things:
* Resize the style based on the DPI, its intended purpose from the method name.
* As a side effect, also change the w and h fields... and if done over and over, well, that can cause very bad results indeed.

I would propose to move the side effect to another method that is called outside of the loop instead, e..g, ``resizeBoundsForDPI``. The call would be before the "for" loops in ``BufferedImageLegendGraphicBuilder`` and ``JSONLegendGraphicBuilder``. @ianturton you seem to be the original author of this method (as far as git goes at least), what do you think?

Reply:
I was just looking at that and thinking it looked familiar! 

Looks as if whatever I was working on only ever called that section of code once. So I would pull out the resize as @aaime suggests.